### PR TITLE
feat: packagemanifest api

### PR DESCRIPTION
**Pull request by yashvardhan-kukreja**
would unblock the package-loader and packagemanifest controller.

Signed-off-by: Yashvardhan Kukreja <ykukreja@redhat.com>

### CODE REVIEW

File: apis/manifests/v1alpha1/packagemanifest_types.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	Name  string `json:"name"`
+	Class string `json:"class,omitempty"`
+}
+
+// PackageList contains a list of Packages.
+type PackageList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []PackageManifest `json:"items"`
+}
```
Comment (thetechnick) [inline apis/manifests/v1alpha1/packagemanifest_types.go:38]: Needs to be registered into manifests/v1alpha1 scheme:
```
func init() {
	SchemeBuilder.Register(&PackageManifest{})
}
```

**Review by thetechnick — CHANGES_REQUESTED** [review]
Please also include a /apis/manifests/manifest.go file to aggregate all versions (Check /apis/core/core.go as example) and add this new API group to `/apis/apis.go`

File: apis/manifests/v1alpha1/packagemanifest_types.go (hunk 1 above)
Comment (yashvardhan-kukreja) [reply apis/manifests/v1alpha1/packagemanifest_types.go:38]: I actually explicitly removed this but just curious why do we need to have its schemeBuilder? We aren't registering this type (PackageManifest) with the Kubernetes API right or is there some other purpose for which we wanna register the Go type to its GVK.

File: apis/manifests/v1alpha1/packagemanifest_types.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	pkoapisv1alpha1 "package-operator.run/apis/core/v1alpha1"
+)
+
+// Cloned from "k8s.io/apimachinery/pkg/apis/meta/v1" so as to not make controller-gen release crds from it
+type TypeMeta metav1.TypeMeta
+
+type PackageManifest struct {
+	TypeMeta `json:",inline"`
+
```
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply apis/manifests/v1alpha1/packagemanifest_types.go:14]: Sure thing, happy to do that. 
Although I couldn't find a way with controller-gen to do it. 
So should I just program it in the magefile's `Generate All` to remove the generated CRDs and static deployments associated with PackageManifest which controller-gen generates?


Regarding generating the `zz_generated.deepcopy.go` , spot on, adding that :)

File: apis/manifests/v1alpha1/groupversion_info.go [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// The package v1alpha1 contains API Schema definitions for the v1alpha1 version of the core Package Operator API group,
+// containing basic building blocks that other auxiliary APIs can build on top of.
```
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply apis/manifests/v1alpha1/groupversion_info.go:2]: whoops

**Comment by yashvardhan-kukreja** [comment]
> Please also include a /apis/manifests/manifest.go file to aggregate all versions (Check /apis/core/core.go as example) and add this new API group to /apis/apis.go

Regarding this Nico, I had to explicitly remove it because, at least to what I noticed in the core/core.go file, the apis.go file was just aggregating the `AddToScheme` of all the "Groups" (core, manifests, etc.).

But in the case of `manifests.package-operator.run`, we are not exporting its scheme. I did that because we don't intend to hook the PackageManifest type as a GVK/resource to the Kubernetes API at anytime, it's just a Kubenetes-compliant format to represent a Package's info. Therefore, due to the absence of Scheme/SchemeBuilder/AddToScheme methods, there was nothing to put in the `apis/apis.go` and `manifests/manifest.go`, therefore, I had to get rid of them. 

Would love to hear your points :)

File: apis/manifests/v1alpha1/packagemanifest_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply apis/manifests/v1alpha1/packagemanifest_types.go:38]: gotcha

File: apis/core/v1alpha1/commonpackage_types.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+	// Failed maps to the fact that the Job created by the Package failed
+	PackageStatusPhaseFailed PackageStatusPhase = "Failed"
+)
+
+type PackageStatusType string
+
+const (
+	PackageStatusConditionTypeJobScheduled PackageStatusType = "JobScheduled"
+	PackageStatusConditionTypeJobSucceeded PackageStatusType = "JobSucceeded"
+	PackageStatusConditionTypeJobFailed    PackageStatusType = "JobFailed"
```
[the remark this answers is not in the document]
Comment (yashvardhan-kukreja) [reply apis/core/v1alpha1/commonpackage_types.go:33]: Sure, agreed.
Gotta exercise my status reporting muscles.

File: apis/manifests/v1alpha1/packagemanifest_types.go (hunk 1 above)
Comment (thetechnick) [reply apis/manifests/v1alpha1/packagemanifest_types.go]: Schemes are not about registering it into the kube-apiserver. It's called `runtime.Scheme` for a reason :)
We use it during runtime to lookup type information and do conversion.

File: apis/manifests/v1alpha1/groupversion_info.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// The package v1alpha1 contains API Schema definitions for the v1alpha1 version of the manifests Package Operator API group,
+// containing basic building blocks that other auxiliary APIs can build on top of.
```
Comment (thetechnick) [inline apis/manifests/v1alpha1/groupversion_info.go:2]:
```suggestion
// The package v1alpha1 contains API Schema definitions for the v1alpha1 version of the manifests API group,
// containing file-based manifests for the packaging infrastructure.
```

**Review by thetechnick — CHANGES_REQUESTED** [review]
A small doc-change, otherwise good to go :)

File: apis/manifests/v1alpha1/packagemanifest_types.go (hunk 1 above)
Comment (yashvardhan-kukreja) [reply apis/manifests/v1alpha1/packagemanifest_types.go]: Understood, thanks for the explanation. Updating :)

**Review by thetechnick — APPROVED** [review]

File: apis/apis.go [hunk 6]
```diff
@@ -7,10 +7,13 @@ import (
 	"package-operator.run/apis/manifests"
 )
 
-// AddToScheme adds all api Resources to the Scheme.
+// AddToSchemes may be used to add all resources defined in the project to a Scheme.
+var AddToSchemes runtime.SchemeBuilder = runtime.SchemeBuilder{
+	core.AddToScheme,
+	manifests.AddToScheme,
+}
+
```
Comment (yashvardhan-kukreja) [inline apis/apis.go:15]: Nice :)

**Pull request merged**